### PR TITLE
Context Parallel integration tests with a transformer layer: BSHD and THD + CP

### PR DESCRIPTION
**Pull request by jomitchellnv**
# Description

This PR creates the following folder
```
TransformerEngine/examples/pytorch/transformer:
├── context_parallel_runner_bshd.py
├── context_parallel_runner_thd.py
├── model.py
├── __pycache__
├── README.md
├── run_context_parallel.sh
├── test_context_parallel_bshd.py
├── test_context_parallel_thd.py
└── utils.py

```
That provides integration tests for the THD and BSHD input formats. The test compares context parallel runs to their non context parallel counterparts. 

Logits, loss, and gradients are compared for both of these scenarios with pass/fail criteria established in the testing files. 

The models architecture consists of an embedding layer and 1 TE.TransformerLayer for simplicity.

Moreover, we also provide utilities for moving data to CP shards when we are using `BSHD` and `SBHD` input formats.

# TODO: Not sure how you want to add this test suite into the broader testing stuff.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:
- Creates a simple transformer and runs integration tests on it for CP and non CP comparisons.
- Improvement to context parallel utility handling, that handles dynamic sequence dimensions during CP sharding, error checking, as well as support for the BSHD and SBHD input formats

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [X] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 1]
```diff
@@ -4132,6 +4132,113 @@ def process_tensor(val):
...
+
+            # Reshape back to original format with reduced sequence dimension
+            val = val.view(-1, *val.shape[2:])
+            return val
+
+        if cp_size > 1:
+            input_ids_padded = process_tensor(input_ids_padded)
+            labels_padded = process_tensor(labels_padded)
+            position_ids_padded = process_tensor(position_ids_padded)
+
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py:4241]: Most of the code is the same - just `seq_dim` is dynamically determined. Could be combined.

File: examples/pytorch/transformer/model.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                cu_seqlens_q=batch["cu_seqlens_q"],
+                cu_seqlens_kv=batch["cu_seqlens_kv"],
+                pad_between_seqs=batch["pad_between_seqs"],
+                max_seqlen_q=batch["max_seqlen_q"],
+                max_seqlen_kv=batch["max_seqlen_kv"],
+            )
+        hidden_states = self.emb_layer_norm_after(hidden_states)
+        logits = self.linear(hidden_states)
+        # Logits of shape [batch_size, seq_length, vocab_size]
+        return logits
```
Comment (sudhakarsingh27) [inline examples/pytorch/transformer/model.py:186]: `SimpleThdModel` and `SimpleBSHDModel` have almost exact same code? Is it possible to create a single class? If not, what'd be the issue?

Comment (sudhakarsingh27) [reply examples/pytorch/transformer/model.py:186]: I'm thinking why `qkv_format` can't be provided as an input argument to differentiate b/w the two workflows

File: examples/pytorch/transformer/utils.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    rank: int = field(default_factory=dist.get_rank)
+    local_rank: int = field(default_factory=lambda: int(os.environ["LOCAL_RANK"]))
+    world_size: int = field(default_factory=dist.get_world_size)
+
+    def is_main_process(self) -> bool:
+        """This is the global rank 0 process, to be used for wandb logging, etc."""
+        return self.rank == 0
+
+
+def get_dummy_data_bshd():
```
Comment (sudhakarsingh27) [inline examples/pytorch/transformer/utils.py:30]: Does this and its `thd` counterpart produce identical data? 
If yes, could we merge these together too so that there's a single source of truth. I remember doing something similar for one of the repros we did a few weeks ago.

**Review by sudhakarsingh27 — COMMENTED** [review]
Thanks for adding the BSHD path as well to this example!

Overall, I feel like there's a lot of duplicated code and we could simplify further along these lines:
1. Have a single model class that can take in both BSHD and THD formats. 
2. Have `context_parallel_runner.py` as the only file which can do both BSHD and THD formats. 
3. The single GPU baseline could remain to be a part of `context_parallel_runner.py` or we could create a separate file for it. The most important part of this example is the distributed context parallel code IMO
4. This will simplify the test files as well into - `test_context_parallel_runner.py`
5. It might look like this for both BSHD and THD formats:
- `context_parallel_runner.py`
- `model.py`
- `run_context_parallel.sh`
- `test_context_paralel.py`
- `utils.py`
6. It's possible to run the example without a separate `run_context_parallel.sh` with something like the following in `context_parallel_runner.py`:

```
def main():
    import os
    world_size = os.environ.get('WORLD_SIZE', 2) # Number of processes
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '29505'

    # Baseline call
    baseline_example_func()

    # CP call
    torch.multiprocessing.spawn(context_parallel_example_func,
                                args=(world_size,),
                                nprocs=world_size,
                                join=True)

if __name__ == "__main__":
    main()
```
This way `context_parallel_runner.py` could be a standalone example file.

Let me know your thoughts.

**Comment by KshitijLakhani** [comment]
@jomitchellnv @sudhakarsingh27 I've removed the 2.12 tag
Please reset it to a more appropriate tag
Thanks !

**Comment by sudhakarsingh27** [comment]
Closing now as discussed. @jomitchellnv to re-open in the future when coming back to this.
